src/main/scala/scala/compat/java8/collectionImpl/Accumulator.scala

```diff
@@ -0,0 +1,344 @@
+  */
+final class Accumulator[A] extends AccumulatorLike[A, Accumulator[A]] { self =>
+  private[java8] var current: Array[AnyRef] = Accumulator.emptyAnyRefArray
+  private[java8] var history: Array[Array[AnyRef]] = Accumulator.emptyAnyRefArrayArray
+  private[java8] var cumul: Array[Long] = Accumulator.emptyLongArray
+  
+  private[java8] def cumulative(i: Int) = cumul(i)
+    
+  private def expand(): Unit = {
+    if (index > 0) {
+      if (hIndex >= history.length) hExpand()
+      history(hIndex) = current
+      cumul(hIndex) = (if (hIndex > 0) cumulative(hIndex-1) else 0) + index
+      hIndex += 1
+    }
+    current = new Array[AnyRef](nextBlockSize)
+    index = 0
+  }
+  
+  private def hExpand(): Unit = {
+    if (hIndex == 0) {
+      history = new Array[Array[AnyRef]](4)
+      cumul = new Array[Long](4)
+    }
+    else {
+      history = java.util.Arrays.copyOf(history, history.length << 1)
+      cumul = java.util.Arrays.copyOf(cumul, cumul.length << 1)
+    }
+  }
+  
+  /** Appends an element to this `Accumulator`. */
+  final def +=(a: A): Unit = {
+    totalSize += 1
+    if (index >= current.length) expand()
+    current(index) = a.asInstanceOf[AnyRef]
+    index += 1
+  }
+  
+  /** Removes all elements from `that` and appends them to this `Accumulator`. */
+  final def drain[A1 <: A](that: Accumulator[A1]): Unit = {
+    var h = 0
+    var prev = 0L
+    var more = true
+    while (more && h < that.hIndex) {
+      val n = (that.cumulative(h) - prev).toInt
+      if (current.length - index >= n) {
+        System.arraycopy(that.history(h), 0, current, index, n)
+        prev = that.cumulative(h)
+        index += n
+        h += 1
+      }
+      else more = false
+    }
+    if (h >= that.hIndex && current.length - index >= that.index) {
+      if (that.index > 0) System.arraycopy(that.current, 0, current, index, that.index)
+      index += that.index
+    }
+    else {
+      val slots = (if (index > 0) 1 else 0) + that.hIndex - h
+      if (hIndex + slots > history.length) {
+        val n = math.max(4, 1 << (32 - java.lang.Integer.numberOfLeadingZeros(1 + hIndex + slots)))
+        history = java.util.Arrays.copyOf(history, n)
+        cumul = java.util.Arrays.copyOf(cumul, n)
+      }
+      var pv = (if (hIndex > 0) cumulative(hIndex-1) else 0L)
+      if (index > 0) {
+        pv += index
+        cumul(hIndex) = pv
+        history(hIndex) = (if (index < (current.length >>> 3) && current.length > 32) java.util.Arrays.copyOf(current, index) else current)
+        hIndex += 1
+      }
+      while (h < that.hIndex) {
+        pv += that.cumulative(h) - prev
+        prev = that.cumulative(h)
+        cumul(hIndex) = pv
+        history(hIndex) = that.history(h)
+        h += 1
+        hIndex += 1
+      }
+      index = that.index
+      current = that.current
+    }
+    totalSize += that.totalSize
+    that.clear
+  }
+  
+  override def clear(): Unit = {
```

Comment:
The only place that calls `drain` is `Accumulator.merger` for use by `Stream.collect`. It is not clear to me from the documentation of `collect` whether implementations are supposed to drop references quickly. I would assume that they generally do though, otherwise all *immutable* accumulation would always retain references to old objects. In particular, the example in the documentation is `stringStream.collect(ArrayList::new, ArrayList::add, ArrayList::addAll)`. This would suffer from the same problem. OTOH, not clearing the RHS Accumulator would leave it in an invalid state. It's probably not worth complicating the API with such a "dangerous" method for the little performance gain that can be expected.